File: CCC/ccc03s3.cpp
#include <bits/stdc++.h>
using namespace std;
int main(){
    cin.tie(nullptr)->sync_with_stdio(false);

    int f, n, m;
    cin>>f>>n>>m;
    vector<vector<char>>arr(n+1, vector<char>(m+1));
    vector<vector<bool>>vis(n+1, vector<bool>(m+1));
    vector<int>changeX{1,-1,0,0};
    vector<int>changeY{0,0,-1,1};
    queue<int>q;
    vector<int>roomSize;
    int sum =0;
    int ans = 0;
    for(int i =1;i<=n;i++){
        string s; cin>>s;
        for(int j=1;j<=m;j++){
            arr[i][j] = s[j-1];
        }
    }
    for(int i =1;i<=n;i++){
        for(int j=1;j<=m;j++){
            if(arr[i][j]!='I' && !vis[i][j]){
                int roomCount = 1;
                vis[i][j] = true;
                q.push(i*100+j);
                while(!q.empty()){
                    int x = q.front()/100; int y = q.front()%100; q.pop();
                    for(int k=0;k<4;k++){
                        int newx = x + changeX[k];
                        int newy = y + changeY[k];
                        if(newx>0 && newx<=n &&newy>0 && newy<=m){
                            if(!vis[newx][newy] && arr[newx][newy]!='I'){
                                vis[newx][newy] = true;
                                roomCount++;
                                q.push(newx*100+newy);
                            }
                        }
                    }
                }
                sum+= roomCount;
                roomSize.push_back(roomCount);
            }
        }
    }
    sort(roomSize.begin(), roomSize.end(), greater<int>());

    for(int i =0;i<roomSize.size();i++){
        if(f-roomSize.at(i) <0) break;
        else if(f - roomSize.at(i)>=0){
            ans ++;
            f -=roomSize.at(i);
        }
    }
    if(ans==1) cout<<"1 room, "<<f<<" square metre(s) left over";
    else cout<<ans<<" rooms, "<<f<<" square metre(s) left over";

    return 0;
}

File: CCC/ccc03s3.java
import java.util.*;
import java.io.*;
public class floorplan{
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static PrintWriter pr = new PrintWriter(new BufferedWriter(new OutputStreamWriter(System.out)));
    static StringTokenizer st;
    public static void main(String[]args)throws IOException {
        
        
        int flooring = readInt();
        int n = readInt();
        int m = readInt();
        ArrayList<Integer>roomCords = new ArrayList();
        ArrayList<Integer>roomSizes = new ArrayList();
        char[][]array = new char[n+1][m+1];
        int filled = 0;

        for(int i = 1; i<=n; i++){
            String s = readLine();
            for(int j = 1; j<=m; j++){
                array[i][j] = s.charAt(j-1);
                if(array[i][j] == '.')
                    roomCords.add(i*100 + j);
            }
        }

        while(!roomCords.isEmpty()){

            int start = roomCords.get(0);
            roomCords.remove(0);
            int count = 1;
            boolean[][]vis = new boolean[n+1][m+1];
            Queue<Integer>q = new LinkedList();
            int[] changex = {0,0,-1,1};
            int[] changey = {1,-1,0,0};
            vis[start/100][start%100] = true;
            q.add(start);

            while(!q.isEmpty()){
                int cur = q.poll();
                int x = cur/100;
                int y = cur%100;

                for(int i = 0; i<4; i++){
                    int newx = x + changex[i];
                    int newy = y + changey[i];
                    if(newx>0 && newx<=n && newy>0 && newy<=m){
                        if(!vis[newx][newy] && array[newx][newy]!='I'){
                            vis[newx][newy] = true;
                            count++;
                            q.add(newx*100+newy);
                            roomCords.remove(roomCords.indexOf(newx*100+newy));
                        }
                    }
                }


            }
            roomSizes.add(count);

        }
        Collections.sort(roomSizes);


        for(int i = roomSizes.size()-1; i>=0; i--){
            if(flooring>=roomSizes.get(i)){
                flooring -= roomSizes.get(i);
                filled++;
            }
            else
                break;
            
        }
        

        if (filled == 1){
            System.out.println("1 room, " + flooring + " square metre(s) left over");
        }
        else    
            System.out.println(filled + " rooms, " + flooring + " square metre(s) left over");
    }
    static String next () throws IOException {
        while (st == null || !st.hasMoreTokens())
            st = new StringTokenizer(br.readLine().trim());
        return st.nextToken();
    }
    static long readLong () throws IOException {
        return Long.parseLong(next());
    }
    static int readInt () throws IOException {
        return Integer.parseInt(next());
    }
    static double readDouble () throws IOException {
        return Double.parseDouble(next());
    }
    static char readChar () throws IOException {
        return next().charAt(0);
    }
    static String readLine () throws IOException {
        return br.readLine().trim();
    }
}